cluster/src/main/java/com/hartwig/pipeline/metadata/BiopsyMetadataApi.java

```diff
@@ -62,6 +85,49 @@ public void start() {
 
     @Override
     public void complete(final PipelineState state, final SomaticRunMetadata metadata) {
-        // do nothing
+        if (state.status() != PipelineStatus.FAILED) {
+            List<AddDatatype> addDatatypes =
+                    state.stageOutputs().stream().map(StageOutput::datatypes).flatMap(List::stream).collect(Collectors.toList());
+            SampleSet set = OnlyOne.of(setApi.list(metadata.set(), null), SampleSet.class);
+            ImmutablePipelineStaged.Builder stagedEventBuilder = ImmutablePipelineStaged.builder()
+                    .type(PipelineStaged.Type.DNA)
+                    .version(Versions.pipelineMajorMinorVersion())
+                    .sample(metadata.tumor().sampleName())
+                    .setId(set.getId());
+            OutputIterator.from(blob -> {
+                Optional<DataType> dataType =
+                        addDatatypes.stream().filter(d -> blob.getName().endsWith(d.path())).map(AddDatatype::dataType).findFirst();
+                stagedEventBuilder.addBlobs(root(metadata.tumor().sampleName(),
+                        metadata.reference().sampleName(),
+                        blob.getName(),
+                        PipelineOutputBlob.builder()
+                                .datatype(dataType.map(Object::toString))
+                                .barcode(metadata.barcode())
+                                .bucket(blob.getBucket())
+                                .filesize(blob.getSize())
+                                .hash(MD5s.convertMd5ToSbpFormat(blob.getMd5()))).build());
+            }, sourceBucket).iterate(metadata);
+            stagedEventBuilder.build().publish(publisher, objectMapper);
+        }
+    }
+
+    private static ImmutablePipelineOutputBlob.Builder root(final String tumorSample, final String refSample, final String blobName,
```

Comment:
Two things I don't like about this method, but for which I don't have great solutions:

* It mutates the state of the builder it is passed in, then returns it. In practice you're creating the builder inline in the caller so I understand why and also it maybe isn't a big problem, just don't love the way it reads.
* Don't love the name. Maybe `pathComponents` or `splitPath`? I dunno.

Reply:
agree on both, i'll see what i can do to improve